File: src/src_platform_uwp/fmg/uwp/mosaic/wbmp/MosaicCanvasView.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Imaging;
using fmg.common.geom;
using fmg.core.mosaic;
using fmg.core.mosaic.cells;
using fmg.uwp.img.wbmp;
using fmg.uwp.utils.wbmp;

namespace fmg.uwp.mosaic.wbmp {

   /// <summary> MVC: view. UWP WriteableBitmap implementation over control <see cref="Canvas"/> </summary>
   public class MosaicCanvasView : MosaicView<Canvas, WriteableBitmap, MosaicDrawModel<WriteableBitmap>> {

      class InnerView : MosaicViewWBmp<WriteableBitmap, MosaicDrawModel<WriteableBitmap>> {

         private readonly MosaicCanvasView _owner;

         public InnerView(MosaicCanvasView self)
            : base(self.Model) {
            _owner = self;
         }

         public override void Draw(IEnumerable<BaseCell> modifiedCells) {
            //Draw(modifiedCells, null, true);
            throw new InvalidOperationException();
         }

         public void Draw(IEnumerable<BaseCell> modifiedCells, RectDouble clipRegion, bool drawBk) {
            base.Draw(modifiedCells, clipRegion, drawBk);
         }

      }

      private InnerView _innerView;
      private Canvas _control;
      private Flag.Controller _imgFlag = new Flag.Controller();
      private Mine.Controller _imgMine = new Mine.Controller();

      public MosaicCanvasView()
         : base(new MosaicDrawModel<WriteableBitmap>())
      {
         _innerView = new InnerView(this);
         ChangeSizeImagesMineFlag();
      }

      protected override Canvas CreateImage() {
         // will return once created window
         return GetControl();
      }

      public Canvas GetControl() {
         if (_control == null) {
            _control = new Canvas();

            var imgControl = new Image {
               Stretch = Stretch.None
            };
            imgControl.SetBinding(Windows.UI.Xaml.Controls.Image.SourceProperty, new Binding {
               Source = Image,
               Path = new PropertyPath(nameof(Image)),
               Mode = BindingMode.OneWay
            });

            _control.Children.Add(imgControl);
         }
         return _control;
      }

      public override void Draw(IEnumerable<BaseCell> modifiedCells) {
         Canvas control = GetControl();
         RectDouble rc = new RectDouble(0,0, control.Width, control.Height);
         _innerView.Draw(modifiedCells, rc, true);
      }

      public override void Invalidate() {
         base.Invalidate();
         //AsyncRunner.InvokeFromUiLater(() => {
         //      var img = Image; // implicit call Draw() -> DrawBegin() -> this.Draw(...)
         //}, CoreDispatcherPriority.High);
      }

      protected override void OnPropertyModelChanged(object sender, PropertyChangedEventArgs ev) {
         base.OnPropertyModelChanged(sender, ev);
         switch (ev.PropertyName) {
         case nameof(MosaicGameModel.MosaicType):
         case nameof(MosaicGameModel.Area):
            ChangeSizeImagesMineFlag();
            break;
         }
      }

      /** переустанавливаю заного размер мины/флага для мозаики */
      protected void ChangeSizeImagesMineFlag() {
         MosaicDrawModel<WriteableBitmap> model = Model;
         int sq = (int)model.CellAttr.GetSq(model.PenBorder.Width);
         if (sq <= 0) {
            System.Diagnostics.Debug.WriteLine("Error: too thick pen! There is no area for displaying the flag/mine image...");
            sq = 3; // ат балды...
         }

         const int max = 30;
         if (sq > max) {
            _imgFlag.Model.SetSize(sq);
            _imgMine.Model.SetSize(sq);
            model.ImgFlag = _imgFlag.Image;
            model.ImgMine = _imgMine.Image;
         } else {
            _imgFlag.Model.SetSize(max);
            model.ImgFlag = ImgUtils.Zoom(_imgFlag.Image, sq, sq);
            _imgMine.Model.SetSize(max);
            model.ImgMine = ImgUtils.Zoom(_imgMine.Image, sq, sq);
         }
      }

      protected override void Disposing() {
         _innerView.Dispose();
         Model.Dispose();
         base.Disposing();
         _control = null;
         _imgFlag.Dispose();
         _imgMine.Dispose();
      }

   }

}

File: src/src_platform_jfx/fmg/jfx/mosaic/MosaicCanvasView.java
package fmg.jfx.mosaic;

import java.beans.PropertyChangeEvent;
import java.util.Collection;

import javafx.scene.canvas.Canvas;
import javafx.scene.image.Image;

import fmg.common.geom.RectDouble;
import fmg.common.geom.SizeDouble;
import fmg.core.mosaic.MosaicDrawModel;
import fmg.core.mosaic.MosaicGameModel;
import fmg.core.mosaic.cells.BaseCell;
import fmg.core.mosaic.cells.Square1;
import fmg.jfx.img.CanvasJfx;
import fmg.jfx.img.Flag;
import fmg.jfx.img.Mine;
import fmg.jfx.utils.ImgUtils;

/** MVC: view. JavaFX implementation over node-control {@link Canvas} */
public class MosaicCanvasView extends MosaicJfxView<Canvas, Image, MosaicDrawModel<Image>> {

    private CanvasJfx _canvas = new CanvasJfx(this);
    private Flag.ImageJfxController _imgFlag = new Flag.ImageJfxController();
    private Mine.ImageJfxController _imgMine = new Mine.ImageJfxController();

    public MosaicCanvasView() {
        super(new MosaicDrawModel<Image>());
        changeSizeImagesMineFlag();
    }

    @Override
    protected Canvas createImage() { return _canvas.create(); }

    @Override
    protected void drawModified(Collection<BaseCell> modifiedCells) {
        if (modifiedCells == null) {
            drawJfx(_canvas.getGraphics(), getModel().getMatrix(), true);
            return;
        }
        if (modifiedCells.iterator().next() instanceof Square1) { // optimize
            drawJfx(_canvas.getGraphics(), modifiedCells, false);
            return;
        }
        double minX=0, minY=0, maxX=0, maxY=0;
        boolean first = true;
        for (BaseCell cell : modifiedCells) {
            RectDouble rc = cell.getRcOuter();
            if (first) {
                first = false;
                minX = rc.x;
                minY = rc.y;
                maxX = rc.right();
                maxY = rc.bottom();
            } else {
                minX = Math.min(minX, rc.x);
                minY = Math.min(minY, rc.y);
                maxX = Math.max(maxX, rc.right());
                maxY = Math.max(maxY, rc.bottom());
            }
        }
        MosaicDrawModel<?> model = getModel();
        SizeDouble offset = model.getMosaicOffset();
        RectDouble rcClip = new RectDouble(minX + offset.width, minY + offset.height, maxX-minX, maxY-minY);
//        if (_DEBUG_DRAW_FLOW)
//            System.out.println("MosaicViewJfx.draw: repaint=" + rcClip);
        drawJfx(_canvas.getGraphics(), toDrawCells(rcClip), false);
    }

    @Override
    protected void onPropertyChanged(PropertyChangeEvent ev) {
        super.onPropertyChanged(ev);
        if (PROPERTY_IMAGE.equals(ev.getPropertyName()))
            getImage(); // implicit call draw() -> drawBegin() -> drawModified() -> drawJfx()
    }

    @Override
    protected void onModelPropertyChanged(PropertyChangeEvent ev) {
        super.onModelPropertyChanged(ev);
        switch (ev.getPropertyName()) {
        case MosaicGameModel.PROPERTY_MOSAIC_TYPE:
        case MosaicGameModel.PROPERTY_AREA:
            changeSizeImagesMineFlag();
            break;
        default:
            // none
        }
    }

    /** переустанавливаю заного размер мины/флага для мозаики */
    protected void changeSizeImagesMineFlag() {
        MosaicDrawModel<Image> model = getModel();
        double sq = model.getCellAttr().getSq(model.getPenBorder().getWidth());
        if (sq <= 0) {
            System.err.println("Error: too thick pen! There is no area for displaying the flag/mine image...");
            sq = 3; // ат балды...
        }

        final int max = 30;
        if (sq > max) {
            _imgFlag.getModel().setSize(new SizeDouble(sq, sq));
            _imgMine.getModel().setSize(new SizeDouble(sq, sq));
            model.setImgFlag(_imgFlag.getImage());
            model.setImgMine(_imgMine.getImage());
        } else {
            _imgFlag.getModel().setSize(new SizeDouble(max, max));
            _imgMine.getModel().setSize(new SizeDouble(max, max));
            model.setImgFlag(ImgUtils.zoom(_imgFlag.getImage(), sq, sq));
            model.setImgMine(ImgUtils.zoom(_imgMine.getImage(), sq, sq));
        }
    }

    @Override
    public void close() {
        super.close();
        getModel().close();
        _canvas = null;
        _imgFlag.close();
        _imgMine.close();
    }

}
